### PR TITLE
Make sure ReferenceCountingSegment.decrement() is invoked correctly when useCache=true

### LINKED ISSUE #8291: ReferenceCountedSegment is acquired too many times
A error when a timeseries query is sent to Druid with high concurrency.

### Affected Version

0.14.0-incubating

### Description

```
org.apache.druid.query.ChainedExecutionQueryRunner - Exception with one of the sequences!

java.lang.IllegalStateException: Attempt to register more than 65535 parties for org.apache.druid.segment.ReferenceCountingSegment$1@50149526[phase = 0 parties = 65535 arrived = 0]

       at java.util.concurrent.Phaser.doRegister(Phaser.java:438) ~[?:1.8.0_151]

       at java.util.concurrent.Phaser.register(Phaser.java:591) ~[?:1.8.0_151]

       at org.apache.druid.segment.ReferenceCountingSegment.increment(ReferenceCountingSegment.java:124) ~[druid-processing-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.segment.realtime.FireHydrant.getIncrementedSegment(FireHydrant.java:76) ~[druid-server-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.segment.realtime.FireHydrant.getAndIncrementSegment(FireHydrant.java:136) ~[druid-server-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.segment.realtime.appenderator.SinkQuerySegmentWalker$3$1.apply(SinkQuerySegmentWalker.java:227) ~[druid-server-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.segment.realtime.appenderator.SinkQuerySegmentWalker$3$1.apply(SinkQuerySegmentWalker.java:214) ~[druid-server-0.14.0-incubating.jar:0.14.0-incubating]

       at com.google.common.collect.Iterators$8.transform(Iterators.java:794) ~[guava-16.0.1.jar:?]

       at com.google.common.collect.TransformedIterator.next(TransformedIterator.java:48) ~[guava-16.0.1.jar:?]

       at com.google.common.collect.Iterators$3.next(Iterators.java:163) ~[guava-16.0.1.jar:?]

       at com.google.common.collect.TransformedIterator.next(TransformedIterator.java:48) ~[guava-16.0.1.jar:?]

       at com.google.common.collect.Iterators.addAll(Iterators.java:357) ~[guava-16.0.1.jar:?]

       at com.google.common.collect.Lists.newArrayList(Lists.java:147) ~[guava-16.0.1.jar:?]

       at com.google.common.collect.Lists.newArrayList(Lists.java:129) ~[guava-16.0.1.jar:?]

       at org.apache.druid.query.ChainedExecutionQueryRunner$1.make(ChainedExecutionQueryRunner.java:104) ~[druid-processing-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.java.util.common.guava.BaseSequence.accumulate(BaseSequence.java:42) ~[druid-core-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.java.util.common.guava.LazySequence.accumulate(LazySequence.java:40) ~[druid-core-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.java.util.common.guava.WrappingSequence$1.get(WrappingSequence.java:50) ~[druid-core-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.java.util.common.guava.SequenceWrapper.wrap(SequenceWrapper.java:55) ~[druid-core-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.java.util.common.guava.WrappingSequence.accumulate(WrappingSequence.java:45) ~[druid-core-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.java.util.common.guava.LazySequence.accumulate(LazySequence.java:40) ~[druid-core-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.java.util.common.guava.WrappingSequence$1.get(WrappingSequence.java:50) ~[druid-core-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.java.util.common.guava.SequenceWrapper.wrap(SequenceWrapper.java:55) ~[druid-core-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.java.util.common.guava.WrappingSequence.accumulate(WrappingSequence.java:45) ~[druid-core-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.java.util.common.guava.WrappingSequence$1.get(WrappingSequence.java:50) ~[druid-core-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.query.CPUTimeMetricQueryRunner$1.wrap(CPUTimeMetricQueryRunner.java:74) ~[druid-processing-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.java.util.common.guava.WrappingSequence.accumulate(WrappingSequence.java:45) ~[druid-core-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.query.spec.SpecificSegmentQueryRunner$1.accumulate(SpecificSegmentQueryRunner.java:82) ~[druid-processing-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.java.util.common.guava.WrappingSequence$1.get(WrappingSequence.java:50) ~[druid-core-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.query.spec.SpecificSegmentQueryRunner.doNamed(SpecificSegmentQueryRunner.java:168) ~[druid-processing-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.query.spec.SpecificSegmentQueryRunner.access$100(SpecificSegmentQueryRunner.java:45) ~[druid-processing-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.query.spec.SpecificSegmentQueryRunner$2.wrap(SpecificSegmentQueryRunner.java:148) ~[druid-processing-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.java.util.common.guava.WrappingSequence.accumulate(WrappingSequence.java:45) ~[druid-core-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.java.util.common.guava.Sequence.toList(Sequence.java:76) ~[druid-core-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.query.ChainedExecutionQueryRunner$1$1.call(ChainedExecutionQueryRunner.java:124) [druid-processing-0.14.0-incubating.jar:0.14.0-incubating]

       at org.apache.druid.query.ChainedExecutionQueryRunner$1$1.call(ChainedExecutionQueryRunner.java:114) [druid-processing-0.14.0-incubating.jar:0.14.0-incubating]

       at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_151]

       at org.apache.druid.query.PrioritizedListenableFutureTask.run(PrioritizedExecutorService.java:247) [druid-processing-0.14.0-incubating.jar:0.14.0-incubating]

       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_151]

       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_151]

       at java.lang.Thread.run(Thread.java:748) [?:1.8.0_151]
```

### PR DESCRIPTION
**Pull request by pphust**
Fixes #8291

The issue itself and solution has been described carefully in [pr8316](https://github.com/apache/incubator-druid/pull/8316), which is for 0.14.2-incubating.

This PR has been applied into latest 0.15.1-incubating branch locally and we come up with one distribution from it successfully. Then we deploy one new middle manage node to test it. It works well in our 0.14.2 cluster also.

### CODE REVIEW

File: server/src/main/java/org/apache/druid/client/CachingQueryRunner.java [hunk 1]
```diff
@@ -92,6 +96,7 @@ public CachingQueryRunner(
     }
 
     if (useCache) {
+      close();
```
[the remark this answers is not in the document]
Comment (pphust) [reply server/src/main/java/org/apache/druid/client/CachingQueryRunner.java]: I agree with you the `close()` should be wrapped in `if (cachedResult != null)` nor` if (useCache=true) `.
The current fix should meet some problem when useCache=true while populateCache=false. 
It is weekend in China. I will test it again and update you later.

File: server/src/main/java/org/apache/druid/client/CloseableCachingQueryRunner.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+import org.apache.druid.client.cache.CacheConfig;
+import org.apache.druid.client.cache.CachePopulator;
+import org.apache.druid.java.util.common.guava.CloseQuietly;
+import org.apache.druid.query.QueryRunner;
+import org.apache.druid.query.QueryToolChest;
+import org.apache.druid.query.SegmentDescriptor;
+
+import java.io.Closeable;
+
+public class CloseableCachingQueryRunner<T> extends CachingQueryRunner
```
[the remark this answers is not in the document]
Comment (pphust) [reply server/src/main/java/org/apache/druid/client/CloseableCachingQueryRunner.java]: I will add some javadoc for `SinkQuerySegmentWalker` as your suggestion.

File: server/src/main/java/org/apache/druid/client/CachingQueryRunner.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (pphust) [reply server/src/main/java/org/apache/druid/client/CachingQueryRunner.java]: I have tested the duplicate decrement scenario. It seems there is some self-protect mechanism in `decrementOnceCloseable()` of ReferenceCountingSegment. So the program does not crash and just prints WARN message like "close() is called more than once on ReferenceCountingSegment".
Anyway I agree with you so I move `close()` into `cachedResult != null` related statement and test it again. All looks good. I have added javadoc for CloseableCachingQueryRunner also.

File: server/src/main/java/org/apache/druid/segment/realtime/appenderator/SinkQuerySegmentWalker.java [hunk 3]
```diff
@@ -231,7 +231,7 @@ public SegmentDescriptor apply(final PartitionChunk<Sink> chunk)
                                                       // 1) Only use caching if data is immutable
                                                       // 2) Hydrants are not the same between replicas, make sure cache is local
                                                       if (hydrantDefinitelySwapped && cache.isLocal()) {
-                                                        QueryRunner<T> cachingRunner = new CachingQueryRunner<>(
+                                                        QueryRunner<T> cachingRunner = new CloseableCachingQueryRunner<>(
```
Comment (leventov) [inline server/src/main/java/org/apache/druid/segment/realtime/appenderator/SinkQuerySegmentWalker.java]: I think you could achieve the same effect using the following code:
```java
QueryRunner<T> runner;
// 1) Only use caching if data is immutable
// 2) Hydrants are not the same between replicas, make sure cache is local
if (hydrantDefinitelySwapped && cache.isLocal()) {
  runner = CachingQueryRunner<>(...);
} else {
  runner = factory.createRunner(segment.lhs);
}
runner = QueryRunnerHelper.makeClosingQueryRunner(runner, segment.rhs);
return new Pair<>(segment.lhs.getDataInterval(), runner);
```

Thus new class `CloseableCachingQueryRunner` is not needed.

Please also:
 - Rename `segment` to something like `segmentAndCloseable`
 - Reduce the level of nesting by extracting variables, such as `FunctionalIterable.create(specs).transform(...)` as `perSegmentRunners` variable.

Comment (pphust) [reply server/src/main/java/org/apache/druid/segment/realtime/appenderator/SinkQuerySegmentWalker.java]: It looks a good idea and I have modified the code correspondingly.
I am not very sure about "Reduce the level of nesting". I think at least we need 2-level cycles in code. One for each sink/segment in time line and the other for each hydrant in sink. So I just change FunctionalIterable.create(specs).transform() to Iterables.transform(). Is it what you want?
The new solved has been cherry-picked into 0.15.1-incubating branch locally and it works well at our cluster.

**Comment by leventov** [comment]
@pphust thanks a lot for contribution!

**Comment by pphust** [comment]
@leventov my pleasure to do something for Druid. 
I have closed the duplicated [PR8316](https://github.com/apache/incubator-druid/pull/8316).

**Comment by licl2014** [comment]
we met same problem even `useCache=false`

**Pull request merged**